google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -321,50 +417,473 @@ private boolean isAlive() {
-    for (PendingModifyAckDeadline modack : ackDeadlineExtensions) {
-      for (List<String> idChunk : Lists.partition(modack.ackIds, MAX_PER_REQUEST_CHANGES)) {
-        ApiFuture<Empty> future =
-            stub.modifyAckDeadlineCallable()
-                .futureCall(
-                    ModifyAckDeadlineRequest.newBuilder()
-                        .setSubscription(subscription)
-                        .addAllAckIds(idChunk)
-                        .setAckDeadlineSeconds(modack.deadlineExtensionSeconds)
-                        .build());
-        ApiFutures.addCallback(future, loggingCallback, directExecutor());
-        pendingOperations++;
+    Map<ModackWithMessageFuture, SettableApiFuture<Map<String, String>>> futureMap =
+        new HashMap<ModackWithMessageFuture, SettableApiFuture<Map<String, String>>>();
+    for (ModackWithMessageFuture modackWithMessageFuture : modackWithMessageFutures) {
+      List<String> ackIdsInRequest = new ArrayList<String>();
+      ApiFutureCallback<Empty> loggingCallback;
+      for (AckIdMessageFuture ackIdMessageFuture :
+          modackWithMessageFuture.getAckIdMessageFutures()) {
+        ackIdsInRequest.add(ackIdMessageFuture.getAckId());
+      }
+
+      if (receiverWithAckResponse != null || enableExactlyOnceDelivery.get()) {
+        // 1) make a response future
+        // 2) add it to our future map
+        // 3) add it to the logging callback
+        SettableApiFuture<Map<String, String>> responseFuture = SettableApiFuture.create();
+        futureMap.put(modackWithMessageFuture, responseFuture);
+        loggingCallback = getLoggingCallback(responseFuture);
+      } else {
+        loggingCallback = getLoggingCallback();
       }
+      ApiFuture<Empty> future =
+          subscriberStub
+              .modifyAckDeadlineCallable()
+              .futureCall(
+                  ModifyAckDeadlineRequest.newBuilder()
+                      .setSubscription(subscription)
+                      .addAllAckIds(ackIdsInRequest)
+                      .setAckDeadlineSeconds(modackWithMessageFuture.getDeadlineExtensionSeconds())
+                      .build());
+      ApiFutures.addCallback(future, loggingCallback, directExecutor());
+      pendingOperations++;
     }
+    ackOperationsWaiter.incrementPendingCount(pendingOperations);
+    return futureMap;
+  }
 
-    for (List<String> idChunk : Lists.partition(acksToSend, MAX_PER_REQUEST_CHANGES)) {
+  private RetryModacksFailedModacks processModackFutures(
+      Map<ModackWithMessageFuture, SettableApiFuture<Map<String, String>>>
+          modackWithMessageFutureMap) {
+    Map<Integer, ModackWithMessageFuture> modacksToRetryMap =
+        new HashMap<Integer, ModackWithMessageFuture>();
+    Set<String> modackIdsFailed = new HashSet<String>();
+    modackWithMessageFutureMap.forEach(
+        (modackWithMessageFuture, settableApiFuture) -> {
+          try {
+            Map<String, String> metadataMap = settableApiFuture.get();
```

Comment:
I don't think we should be doing this synchronously. I think we should instead add a callback to the future, and then process the responses asynchronously.

Reply:
Discussed offline, updated